src/discovery/Discovery.ts

```diff
@@ -175,299 +197,291 @@ class Discovery {
+                continue;
+              } else {
+                throw e;
               }
             }
-            break;
-          default:
-            never();
+            // With this verified chain, we can link
+            const linkedVertexNodeInfo: NodeInfo = {
+              id: nodesUtils.encodeNodeId(linkedVertexNodeId),
+              chain: linkedVertexChainData,
+            };
+            await this.gestaltGraph.linkNodeAndIdentity(
+              linkedVertexNodeInfo,
+              vertexIdentityInfo,
+            );
+            // Add this vertex to the queue if it is not present
+            if (!this.visitedVertices.has(linkedVertexGK)) {
+              await this.startDiscoveryForVertex(linkedVertexGK);
+            }
+          }
         }
-        this.visitedVertices.add(vertex);
-        await this.removeKeyFromDiscoveryQueue(vertexId);
-      }
+        break;
+      default:
+        never();
     }
+    this.visitedVertices.add(vertex);
   }
 
   /**
-   * Will resolve once the queue has drained
+   * Will resolve once all discovery tasks have finished
    */
   public async waitForDrained(): Promise<void> {
-    await this.queueDrained.p;
+    let promises: Array<Promise<any>>;
+    do {
+      promises = [];
+      for await (const task of this.taskManager.getTasks('asc', false, [
+        this.constructor.name,
+        this.discoverVertexHandlerId,
+      ])) {
+        promises.push(task.promise());
+      }
+      await Promise.all(promises);
+    } while (promises.length > 0);
   }
 
   /**
    * Simple check for whether the Discovery Queue is empty.
    */
   protected async queueIsEmpty(): Promise<boolean> {
-    let nextDiscoveryQueueId: DiscoveryQueueId | undefined;
-    const keyIterator = this.db.iterator(this.discoveryQueueDbPath, {
-      limit: 1,
-      values: false,
-    });
-    for await (const [keyPath] of keyIterator) {
-      const key = keyPath[0] as Buffer;
-      nextDiscoveryQueueId = IdInternal.fromBuffer<DiscoveryQueueId>(key);
+    for await (const _ of this.taskManager.getTasks('asc', true, [
+      this.constructor.name,
+      this.discoverVertexHandlerId,
+    ])) {
+      return false;
     }
-    return nextDiscoveryQueueId == null;
+    return true;
   }
 
   /**
-   * Push a Gestalt Key to the Discovery Queue. This process also unlocks
-   * the queue if it was previously locked (due to being empty)
-   * Will only add the Key if it does not already exist in the queue
+   * Creates a task that to discover a vertex.
+   * Will not create a new task if an existing task for the vertex exists.
    */
-  protected async pushKeyToDiscoveryQueue(
-    gestaltKey: GestaltKey,
-  ): Promise<void> {
-    await this.db.withTransactionF(async (tran) => {
-      const valueIterator = tran.iterator<GestaltKey>(
-        this.discoveryQueueDbPath,
-        { valueAsBuffer: false },
+  protected async startDiscoveryForVertex(
+    vertex: GestaltKey,
+    tran?: DBTransaction,
+  ) {
```

Comment:
Calling it `startDiscoveryForVertex` seems bad, maybe `queueDiscoveryForVertex`?

Reply:
Why not re-use terminology from nodes? Should this be schedule...

Reply:
I've called it `scheduleDiscoveryForVertex` for now. I don't really see an analogue name in nodes.